litex_boards/targets/zcu104.py

```diff
@@ -0,0 +1,107 @@
+        self.clock_domains.cd_sys    = ClockDomain()
+        self.clock_domains.cd_sys4x  = ClockDomain(reset_less=True)
+        self.clock_domains.cd_pll4x  = ClockDomain(reset_less=True)
+        self.clock_domains.cd_clk200 = ClockDomain()
+        self.clock_domains.cd_ic     = ClockDomain()
+
+        self.submodules.pll = pll = USMMCM(speedgrade=-2)
+        pll.register_clkin(platform.request("clk125"), 125e6)
+
+        pll.create_clkout(self.cd_pll4x, sys_clk_freq*4, buf=None, with_reset=False)
+        pll.create_clkout(self.cd_clk200, 200e6, with_reset=False)
+
+        self.specials += [
+            Instance("BUFGCE_DIV", name="main_bufgce_div",
+                p_BUFGCE_DIVIDE=4,
+                i_CE=1, i_I=self.cd_pll4x.clk, o_O=self.cd_sys.clk),
+            Instance("BUFGCE", name="main_bufgce",
+                i_CE=1, i_I=self.cd_pll4x.clk, o_O=self.cd_sys4x.clk),
+            AsyncResetSynchronizer(self.cd_clk200, ~pll.locked),
+        ]
+
+        ic_reset_counter = Signal(max=64, reset=63)
+        ic_reset = Signal(reset=1)
+        self.sync.clk200 += \
+            If(ic_reset_counter != 0,
+                ic_reset_counter.eq(ic_reset_counter - 1)
+            ).Else(
+                ic_reset.eq(0)
+            )
+        ic_rdy = Signal()
+        ic_rdy_counter = Signal(max=64, reset=63)
+        self.cd_sys.rst.reset = 1
+        self.comb += self.cd_ic.clk.eq(self.cd_sys.clk)
+        self.sync.ic += [
+            If(ic_rdy,
+                If(ic_rdy_counter != 0,
+                    ic_rdy_counter.eq(ic_rdy_counter - 1)
+                ).Else(
+                    self.cd_sys.rst.eq(0)
+                )
+            )
+        ]
+        self.specials += [
+            Instance("IDELAYCTRL", p_SIM_DEVICE="ULTRASCALE",
+                     i_REFCLK=ClockSignal("clk200"), i_RST=ic_reset,
+                     o_RDY=ic_rdy),
+            AsyncResetSynchronizer(self.cd_ic, ic_reset)
+        ]
+
+class BaseSoC(SoCSDRAM):
+    def __init__(self, sys_clk_freq=int(125e6), **kwargs):
+        platform = zcu104.Platform()
+
+        # SoCSDRAM ---------------------------------------------------------------------------------
+        SoCSDRAM.__init__(self, platform, clk_freq=sys_clk_freq, **kwargs)
+
+        # CRG --------------------------------------------------------------------------------------
+        self.submodules.crg = _CRG(platform, sys_clk_freq)
+
+        # DDR4 SDRAM -------------------------------------------------------------------------------
+        if not self.integrated_main_ram_size:
+            self.submodules.ddrphy = usddrphy.USDDRPHY(platform.request("ddram"),
+                memtype      = "DDR4",
+                sim_device   = "ULTRASCALE_PLUS",
+                sys_clk_freq = sys_clk_freq)
+            self.add_csr("ddrphy")
+            self.add_constant("USDDRPHY", None)
+            # FIXME: Use correct DDR4 module
+            sdram_module = EDY4016A(sys_clk_freq, "1:4")
```

Comment:
I did play with LiteX on the ZCU104 a while ago, never got round to something upstreamable in the end due to other priorities. But this DDR config worked https://github.com/daveshah1/litedram/commit/f08cfe6d11457b79cac4676ff7044657a1d19fd5#diff-bc9fdc9a0f1aa798161b2333f66733b7R450 and this was my board file https://github.com/daveshah1/litex-boards/blob/6d7eb0df09d7c886fa93b26f6d763dc78082408c/litex_boards/community/platforms/zcu104.py